src/plugins/janus_sip.c

```diff
@@ -7392,6 +7419,56 @@ gpointer janus_sip_sofia_thread(gpointer user_data) {
 	return NULL;
 }
 
+/* check peer rtp has rfc2833 and push event */
+void janus_sip_check_rfc2833(janus_sip_session *session, char *buffer) {
+	janus_rtp_header *rtp_header = (janus_rtp_header *)buffer;
+	if (rtp_header->type != JANUS_RTP_RFC2833_DEFAULT_PAYLOAD_TYPE)
+		return;
+	char *payload_buffer = buffer + JANUS_RTP_HEADER_LENGTH + rtp_header->csrccount * sizeof(uint32_t);
+	janus_rtp_rfc2833_payload *rfc2833_payload = (janus_rtp_rfc2833_payload *)payload_buffer;
+	uint16_t duration = ntohs(rfc2833_payload->duration);
+	if (rfc2833_payload->end == 0)
+		return;
+
+	/* set up cache to avoid duplication */
+	if(session->dtmf_cache != NULL && session->dtmf_cache->dtmf_event_id == rfc2833_payload->event && session->dtmf_cache->timestamp == rtp_header->timestamp)
+		return;
+	if(session->dtmf_cache != NULL){
+		g_free(session->dtmf_cache);
+		session->dtmf_cache = NULL;
+	}
+	session->dtmf_cache = g_malloc0(sizeof(janus_sip_dtmf_cache));
+	session->dtmf_cache->dtmf_event_id = rfc2833_payload->event;
+	session->dtmf_cache->timestamp = rtp_header->timestamp;
```

Comment:
I don't understand why it's called a cache, and why it's allocated to be stored, especially since you're firing an event right away, which means you won't actually need or read it later on.  I see you mention avoiding duplication, but in that case you don't need an allocation: just add the struct as it is there, and update it's fields. Per code style, it's more apt to call it "latest_dtmf", since cache is misleading as a name.

Reply:
I'm sorry for my poor English. I may not understand what you mean. In my last commit, I changed the variable name and the parameter name in session to ”latest_dtmf“.